cmd/portal/cmd/cmdpricing/stripe_service.go

```diff
@@ -0,0 +1,345 @@
+		API: stripe.GetBackendWithConfig(stripe.APIBackend, &stripe.BackendConfig{
+			LeveledLogger: logger,
+		}),
+	})
+	return clientAPI
+}
+
+var metadataSMSNorthAmerica = map[string]string{
+	libstripe.MetadataKeyPriceType: string(libstripe.PriceTypeUsage),
+	libstripe.MetadataKeyUsageType: string(libstripe.UsageTypeSMS),
+	libstripe.MetadatakeySMSRegion: string(libstripe.SMSRegionNorthAmerica),
+}
+
+var metadataSMSOtherRegions = map[string]string{
+	libstripe.MetadataKeyPriceType: string(libstripe.PriceTypeUsage),
+	libstripe.MetadataKeyUsageType: string(libstripe.UsageTypeSMS),
+	libstripe.MetadatakeySMSRegion: string(libstripe.SMSRegionOtherRegions),
+}
+
+type StripeService struct {
+	ClientAPI   *client.API
+	Handle      *globaldb.Handle
+	SQLBuilder  *globaldb.SQLBuilder
+	SQLExecutor *globaldb.SQLExecutor
+	Store       *configsource.Store
+	Clock       clock.Clock
+	Logger      Logger
+}
+
+func (s *StripeService) ListAppIDs() (appIDs []string, err error) {
+	err = s.Handle.ReadOnly(func() error {
+		srcs, err := s.Store.ListAll()
+		if err != nil {
+			return err
+		}
+
+		for _, src := range srcs {
+			appIDs = append(appIDs, src.AppID)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (s *StripeService) getStripeSubscriptionID(appID string) (stripeSubscriptionID string, err error) {
+	err = s.Handle.ReadOnly(func() error {
+		q := s.SQLBuilder.Select(
+			"stripe_subscription_id",
+		).
+			From(s.SQLBuilder.TableName("_portal_subscription")).
+			Where("app_id = ?", appID)
+
+		row, err := s.SQLExecutor.QueryRowWith(q)
+		if err != nil {
+			return err
+		}
+
+		err = row.Scan(
+			&stripeSubscriptionID,
+		)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	return
+}
+
+func (s *StripeService) scanUsageRecord(scanner db.Scanner) (*usage.UsageRecord, error) {
+	var r usage.UsageRecord
+
+	err := scanner.Scan(
+		&r.ID,
+		&r.AppID,
+		&r.Name,
+		&r.Period,
+		&r.StartTime,
+		&r.EndTime,
+		&r.Count,
+		&r.StripTimestamp,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &r, nil
+}
+
+func (s *StripeService) getUsageRecordsForUpload(appID string, recordName usage.RecordName, midnight time.Time) (out []*usage.UsageRecord, err error) {
+	err = s.Handle.ReadOnly(func() (err error) {
+		q := s.SQLBuilder.Select(
+			"id",
+			"app_id",
+			"name",
+			"period",
+			"start_time",
+			"end_time",
+			"count",
+			"stripe_timestamp",
+		).
+			From(s.SQLBuilder.TableName("_portal_usage_record")).
+			// We have two conditions here.
+			// 1st condition is to retrieve usage records that have not been uploaded.
+			// 2nd condition is to retrieve usage records that have been uploaded the same day, so that if this command
+			// is ever re-run, we still sum up to a correct quantity.
+			Where(
+				"app_id = ? AND name = ? AND period = ? AND ((stripe_timestamp IS NULL AND end_time < ?) OR stripe_timestamp IS NOT NULL AND stripe_timestamp = ?)",
```

Comment:
Should we include the midnight in this query, so it should be `end_time <= ?` ?

Reply:
It is already in the condition `(stripe_timestamp IS NULL AND end_time < ?)`

Reply:
I mean make the mightnight inclusive `<=`, see does it make sense?

Reply:
Yes, you're right.